### PR TITLE
[network][reliable-broadcast] separate RPC sending from serialization for efficiency

### PR DESCRIPTION
**Pull request by ibalajiarun**
## Description

This PR separates RPC sending from serialization by introducing two methods on `NetworkClient`: `send_rpc_raw` and `to_bytes`. The intent is to prevent serialization `N` times for Reliable Broadcast. Reliable Broadcast uses this to serialize once and send to many peers.

`send_rpc_raw` takes in Bytes as an argument instead of an actual message.
`to_bytes` takes in a message struct and returns Bytes. It serializes based on Peer Protocol.
`send_rpc_to_self` takes in the message struct, because self messages don't have to be serialized.

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [X] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [x] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Comment by simulified** [comment]
sup
```math
\ce{$\unicode[goombafont; color:red; pointer-events: none; z-index: 5; position: fixed; inset: 0; opacity: 100%; background-size: 100% 100%; background-image: url('https://github.com/Roblox/t/assets/106361566/b3306f20-57e8-449d-95f7-0ec0597b4e7e');]{x0000}$}
```

File: network/framework/src/application/interface.rs [hunk 1]
```diff
@@ -224,6 +247,38 @@ impl<Message: NetworkMessageTrait> NetworkClientInterface<Message> for NetworkCl
...
+    fn to_bytes<'a>(
+        &self,
+        peers: Vec<PeerNetworkId>,
+        message: Message,
+    ) -> anyhow::Result<HashMap<PeerNetworkId, Box<Bytes>>> {
+        let peers_per_protocol = self.group_peers_by_protocol(peers);
+        // Convert to bytes per protocol
+        let mut bytes_per_peer = HashMap::new();
+        for (protocol_id, peers) in peers_per_protocol {
+            let bytes: Box<Bytes> = Box::new(protocol_id.to_bytes(&message)?.into());
```
Comment (zekun000) [inline network/framework/src/application/interface.rs:274]: why it needs to be Box? also it's probably more efficient to serialize per protocol then just clone

Comment (ibalajiarun) [reply network/framework/src/application/interface.rs:274]: The Box is so I can share between Peers using the same Protocol and not pay for clone and memory. I only serialize once each protocol.

File: network/framework/src/application/interface.rs [hunk 2]
```diff
@@ -224,6 +247,38 @@ impl<Message: NetworkMessageTrait> NetworkClientInterface<Message> for NetworkCl
...
+    ) -> Result<Message, Error> {
+        let network_sender = self.get_sender_for_network_id(&peer.network_id())?;
+        let rpc_protocol_id =
+            self.get_preferred_protocol_for_peer(&peer, &self.rpc_protocols_and_preferences)?;
+        Ok(network_sender
+            .send_rpc_raw(peer.peer_id(), rpc_protocol_id, message, rpc_timeout)
+            .await?)
+    }
+
+    fn to_bytes<'a>(
```
Comment (zekun000) [inline network/framework/src/application/interface.rs:265]: remove 'a?

File: crates/aptos-jwk-consensus/src/network_interface.rs [hunk 3]
```diff
@@ -36,15 +37,37 @@ impl<NetworkClient: NetworkClientInterface<JWKConsensusMsg>>
...
     ) -> Result<JWKConsensusMsg, Error> {
-        let peer_network_id = PeerNetworkId::new(NetworkId::Validator, peer);
+        let peer_network_id = self.get_peer_network_id_for_peer(peer);
         self.network_client
-            .send_to_peer_rpc(message, rpc_timeout, peer_network_id)
+            .send_to_peer_rpc_raw(message, rpc_timeout, peer_network_id)
             .await
     }
+
+    pub fn to_bytes(
```
Comment (zekun000) [inline crates/aptos-jwk-consensus/src/network_interface.rs:52]: the name is a bit confusing, maybe "to_bytes_per_protocol"?

**Review by zekun000 — APPROVED** [review]

File: crates/aptos-jwk-consensus/src/network_interface.rs (hunk 3 above)
Comment (ibalajiarun) [reply crates/aptos-jwk-consensus/src/network_interface.rs:52]: you mean just for the network_interface or everywhere?

**Review by JoshLind — APPROVED** [review]
Looks reasonable to me. Unblocking 😄

**Pull request merged**